numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
 
     if out is None:
         # no-op
         return buf
 
-    # handle ndarray destination
-    if isinstance(out, np.ndarray):
+    # obtain ndarrays, casting to the same data type
+    buf = ensure_ndarray_from_memory(buf).view('u1')
+    out = ensure_ndarray_from_memory(out).view('u1')
 
-        # view source as destination dtype
-        if isinstance(buf, np.ndarray):
-            buf = buf.view(dtype=out.dtype).reshape(-1, order='A')
-        else:
-            buf = np.frombuffer(buf, dtype=out.dtype)
+    # copy memory
+    np.copyto(out, buf)
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    return out
 
-        # copy via numpy
-        np.copyto(out, buf)
 
-    # handle generic buffer destination
-    else:
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
-        # obtain memoryview of destination
-        dest = memoryview(out)
 
-        # ensure source is 1D
-        if isinstance(buf, np.ndarray):
-            buf = buf.reshape(-1, order='A')
-            # try to match itemsize
-            dtype = 'u%s' % dest.itemsize
-            buf = buf.view(dtype=dtype)
+def ensure_ndarray_from_memory(o, flatten=True):
+    """Convenience function to obtain a numpy ndarray using memory exposed by object `o`,
+    ensuring that no memory copies are made, and that `o` is not an object array.
 
-        # try to copy via memoryview
-        dest[:] = buf
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
 
-    return out
+    Returns
+    -------
+    o : ndarray
 
+    """
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
-    else:
-        arr = np.frombuffer(buf, dtype=dtype)
-    return arr
+    if not isinstance(o, np.ndarray):
 
+        # first try to obtain a memoryview or buffer, needed to ensure that we don't
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
+            # accept objects exposing either old-style or new-style buffer interface
+            try:
+                o = memoryview(o)
+            except TypeError:
+                o = buffer(o)
```

Comment:
Would use `np.getbuffer` here. This behaves the same as `buffer` except it will preserve the same write permissions on the underlying data; whereas, `buffer` forces this to be read-only regardless.

Reply:
Yep, could do. FWIW no codecs require a writable buffer, and it's not immediately obvious what np.getbuffer() is doing without doing some digging, so I had a slight preference for a bare ``buffer()`` call to make it more obvious this is accessing old-style buffers.

Reply:
Is this being used by `memory_copy` though?

As to what NumPy is doing, It's a good question. Though it turns out to be very simple. Here's the [C code]( https://github.com/numpy/numpy/blob/de28edd8f514b82c0524b55f622078d47f479322/numpy/core/src/multiarray/multiarraymodule.c#L3493 ).

Basically it tries to get a writable pointer to the object's data, if it fails it creates a read-only buffer. If it succeeds, it creates a read-write buffer. That's pretty much it.

Reply:
Yep, maybe it is important for ``memory_copy``, good point.

Yeah ``np.getbuffer()`` is simple, but you do have to go trawling the numpy C source code for it. I've spent a fair amount of time today trawling numpy C source code to figure out WTF it's doing, especially inside ``np.array()``, was just thinking to save others the pain :smiley:.

Reply:
Yeah, I feel your pain. Same thing with the (new) buffer protocol support. It's probably worth writing up an issue for NumPy devs requesting clearer documentation about this stuff.

Reply:
Yes some sort of feedback to numpy devs would seem appropriate, not quite sure how to frame it yet though. One thing I find particularly confusing is the circumstances under which ``np.array()`` will use the buffer interface (and share memory) versus when it will not. E.g., if I have a bytes object like:

```
a = b'asdfghjkl'
```

...then the following **will** result in a copy:

```
b = np.array(a, copy=False)
```

...whereas the following **will not** result in a copy:

```
b = np.array(memoryview(a), copy=False)
```

The docstring for ``np.array()`` does not even mention the buffer interface at all, although it does mention the "array interface", but I thought that was a different thing, although I may be confused.

Also, numpy has the ``np.frombuffer()`` function, which you would think is exactly what we're after here, as it is explicitly documented as a way to obtain a numpy array from some object exposing a memory buffer. But in order to use that function, you have to provide a dtype explicitly. Numpy could (should?) be obtaining dtype information from the buffer interface by default IMO. You could potentially work around that by doing something like:

```
m = memoryview(a)
b = np.frombuffer(m, dtype=m.format)
```

...but then I'm not completely clear on whether it's cool to use the struct format codes as a numpy dtype. It seems to work, but I did also find an open issue about incompatibilities (https://github.com/numpy/numpy/issues/9642).